src/Query/ConditionGroup.php

```diff
@@ -0,0 +1,194 @@
+   */
+  public function getConjunction() {
+    return $this->conjunction;
+  }
+
+  /**
+   * Adds a condition.
+   *
+   * @param string|\Drupal\entity\Query\ConditionGroup $field
+   *   Either a condition group (for nested AND/OR conditions), or a
+   *   field name with an optional column name. E.g: 'uid', 'address.locality'.
+   * @param mixed $value
+   *   The value.
+   * @param string $operator
+   *   The operator.
+   *   Possible values: =, <>, <, <=, >, >=, BETWEEN, NOT BETWEEN,
+   *                   IN, NOT IN, IS NULL, IS NOT NULL.
+   *
+   * @return $this
+   */
+  public function addCondition($field, $value = NULL, $operator = NULL) {
+    if ($field instanceof ConditionGroup) {
+      if ($field->count() === 1) {
+        // The condition group only has a single condition, merge it.
+        $this->conditions[] = reset($field->getConditions());
+        $this->addCacheTags($field->getCacheTags());
+        $this->addCacheContexts($field->getCacheContexts());
+        $this->mergeCacheMaxAge($field->getCacheMaxAge());
+      }
+      elseif ($field->count() > 1) {
+        $this->conditions[] = $field;
+      }
+    }
+    else {
+      $this->conditions[] = new Condition($field, $value, $operator);
+    }
+
+    return $this;
+  }
+
+  /**
+   * Gets all conditions and nested condition groups.
+   *
+   * @return \Drupal\entity\Query\Condition[]|\Drupal\entity\Query\ConditionGroup[]
+   *   The conditions, where each one is either a Condition or a nested
+   *   ConditionGroup. Returned by reference, to allow callers to replace
+   *   or remove conditions.
+   */
+  public function &getConditions() {
+    return $this->conditions;
+  }
+
+  /**
+   * Clones the contained conditions when the condition group is cloned.
+   */
+  public function __clone() {
+    foreach ($this->conditions as $i => $condition) {
+      $this->conditions[$i] = clone $condition;
+    }
+  }
+
+  /**
+   * Gets the string representation of the condition group.
+   *
+   * @return string
+   *   The string representation of the condition group.
+   */
+  public function __toString() {
+    // Special case for a single, nested condition group:
+    if (count($this->conditions) == 1) {
+      return (string) reset($this->conditions);
+    }
+    $lines = [];
+    foreach ($this->conditions as $condition) {
+      $lines[] = str_replace("\n", "\n  ", (string) $condition);
+    }
+    return $lines ? "(\n  " . implode("\n    {$this->conjunction}\n  ", $lines) . "\n)" : '';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function count() {
+    return count($this->conditions);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCacheTags() {
+    $tags = $this->cacheTags;
+    foreach ($this->conditions as $condition) {
+      if ($condition instanceof ConditionGroup) {
+        $tags = array_merge($tags, $condition->getCacheTags());
+      }
+    }
+    return Cache::mergeTags($tags, []);
```

Comment:
This must be evaluated for each child every time this method is called because `public function &getConditions()` would permit any pre-computed value to be invalidated 👍